shell/server/proxy.js

```diff
@@ -1118,9 +1118,57 @@ tryProxyRequest = (hostId, req, res) => {
 
     return Promise.resolve(true);
   } else {
-    const isAlreadyOpened = req.headers.cookie && req.headers.cookie.indexOf("sandstorm-sid=") !== -1;
+    const isAlreadyOpened = req.headers.cookie &&
+        req.headers.cookie.indexOf("sandstorm-sid=") !== -1;
+
     return getProxyForHostId(hostId, isAlreadyOpened).then((proxy) => {
       if (proxy) {
+        // Let's do some CSRF defense.
+        const expectedOrigin = PROTOCOL + "//" + req.headers.host;
```

Comment:
The reason `mainUrl` has to be permitted in the `referer` case is because the initial page loaded in the iframe will have the main URL as its referer. No browser sends an `Origin` header in this case.

That said, now that I think of it, I'm not entirely sure if browsers are allowed to send `Origin` in this case. I know they are explicitly not supposed to send an `Origin` header for hyperlinks, but I don't know about iframes. Do you happen to know? Maybe I should allow `mainUrl` as an origin to be safe?